pkg/autoupdate/tuf_autoupdate.go

```diff
@@ -0,0 +1,211 @@
+	filejsonstore "github.com/theupdateframework/go-tuf/client/filejsonstore"
+)
+
+//go:embed assets/tuf-dev/root.json
+var rootJson []byte
+
+const (
+	DefaultTufServer = "https://tuf-devel.kolide.com"
+)
+
+type TufAutoupdater struct {
+	metadataClient  *client.Client
+	mirrorClient    *http.Client
+	mirrorUrl       string
+	binary          string
+	operatingSystem string
+	channel         UpdateChannel
+	checkInterval   time.Duration
+	errorCounter    []int64
+	lock            sync.RWMutex
+	interrupt       chan struct{}
+	logger          log.Logger
+}
+
+type TufAutoupdaterOption func(*TufAutoupdater)
+
+func WithTufLogger(logger log.Logger) TufAutoupdaterOption {
+	return func(ta *TufAutoupdater) {
+		ta.logger = log.With(logger, "component", "tuf_autoupdater")
+	}
+}
+
+func WithChannel(channel UpdateChannel) TufAutoupdaterOption {
+	return func(ta *TufAutoupdater) {
+		ta.channel = channel
+	}
+}
+
+func WithUpdateCheckInterval(checkInterval time.Duration) TufAutoupdaterOption {
+	return func(ta *TufAutoupdater) {
+		ta.checkInterval = checkInterval
+	}
+}
+
+func NewTufAutoupdater(metadataUrl, mirrorUrl, binaryPath, rootDirectory string, opts ...TufAutoupdaterOption) (*TufAutoupdater, error) {
+	binaryName := filepath.Base(binaryPath)
+
+	// Set up the local TUF directory for our TUF client -- a dev repo, to be replaced once we move to production
+	localTufDirectory := filepath.Join(rootDirectory, fmt.Sprintf("%s-tuf-dev", strings.TrimSuffix(binaryName, ".exe")))
+	if err := os.MkdirAll(localTufDirectory, 0750); err != nil {
+		return nil, fmt.Errorf("could not make local TUF directory %s: %w", localTufDirectory, err)
+	}
+
+	// Set up our local store i.e. point to the directory in our filesystem
+	localStore, err := filejsonstore.NewFileJSONStore(localTufDirectory)
+	if err != nil {
+		return nil, fmt.Errorf("could not initialize local TUF store: %w", err)
+	}
+
+	// Set up our remote store i.e. tuf-devel.kolide.com
+	remoteOpts := client.HTTPRemoteOptions{
+		MetadataPath: "/repository",
+	}
+	remoteStore, err := client.HTTPRemoteStore(metadataUrl, &remoteOpts, http.DefaultClient)
+	if err != nil {
+		return nil, fmt.Errorf("could not initialize remote TUF store: %w", err)
+	}
+
+	metadataClient := client.NewClient(localStore, remoteStore)
+	if err := metadataClient.Init(rootJson); err != nil {
+		return nil, fmt.Errorf("failed to initialize TUF client with root JSON: %w", err)
+	}
+
+	ta := &TufAutoupdater{
+		metadataClient:  metadataClient,
+		mirrorClient:    http.DefaultClient,
+		mirrorUrl:       mirrorUrl,
+		binary:          binaryName,
+		operatingSystem: runtime.GOOS,
+		channel:         Stable,
+		interrupt:       make(chan struct{}),
+		checkInterval:   60 * time.Second,
+		errorCounter:    make([]int64, 0), // For now, the error counter is a simple list of timestamps when errors occurred
+		logger:          log.NewNopLogger(),
+	}
+
+	for _, opt := range opts {
+		opt(ta)
+	}
+
```

Comment:
I recognize that this works, but this feels a little out of order to me.

In the `NewTufAutoupdater` function, it creates all the tuf things, then it creates the `TufAutoupdater` struct. But this means that if we want to use the functional argument pattern to change them, it's hard. I've generally reversed these.

Create `ta` first, parts opts, then create the TUF things based on the values in `ta`. (And add new things to `ta`)

Reply:
Sure, I'll update this and handle not using the default http clients at the same time as well 👍